state/modelgeneration.go

```diff
@@ -318,7 +344,7 @@ func (g *Generation) UpdateCharmConfig(appName string, master *Settings, validCh
 					{{"txn-revno", g.doc.TxnRevno}},
 				}}},
 				Update: bson.D{
-					{"$set", bson.D{{"charm-config", g.doc.Config}}},
+					{"$set", bson.D{{"charm-config." + appName, makeItemChanges(newDelta)}}},
```

Comment:
I'm assuming here that if the appName didn't have any value before, this just does the right thing.

Reply:
Yes, works fine from a standing start.